### PR TITLE
rpc: define spec for RPC

### LINKED ISSUE #273: spec: define an RPC spec
## Summary

Write a spec for the JSON RPC. As another client is beginning we should specify the JSON RPC so that both clients have something to implement against instead of going off of what tendermint-go does. 
____

#### For Admin Use

- [x] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

### PR DESCRIPTION
**Pull request by marbar3778**
## Description

This PR decribes the RPC layer of tendermint. For now this takes what is present on the Go implementation and calls it the canonical implementation. After this is merged we will be able to reduce or expand the spec and anything outside the spec is an implementation detail.

closes #273 

this spec does not contain:

- tx_search
- block_search
- dump_consensus_state
- consensus_state
- broadcast_tx_commit
- dial_seeds
- dial_peers

Many of these routes are implementation specific and my not be shared across implementations.

### CODE REVIEW

File: spec/rpc/README.md [hunk 1]
```diff
@@ -0,0 +1,1267 @@
...
+  | Routes                                  | Tendermint-Go | Tendermint-Rs |
+  |-----------------------------------------|:-------------:|:-------------:|
+  | [Subscribe](#subscribe)                 |       ✅       |       ✅       |
+  | [Unsubscribe](#unsubscribe)             |       ✅       |       ✅       |
+  | [UnsubscribeAll](#unsubscribeall)       |       ✅       |       ❌       |
+  | [Health](#health)                       |       ✅       |       ✅       |
+  | [Status](#status)                       |       ✅       |       ✅       |
+  | [NetInfo](#netinfo)                     |       ✅       |       ✅       |
+  | [blockchain](#blockchain)               |       ✅       |       ✅       |
+  | [block](#block)                         |       ✅       |       ✅       |
```
Comment (cmwaters) [inline spec/rpc/README.md:23]: Use capitals for consistency

File: spec/rpc/README.md [hunk 2]
```diff
@@ -0,0 +1,1267 @@
...
+  | JSON-RPC 2.0 |       ✅       |       ✅       |
+  | HTTP         |       ✅       |       ✅       |
+  | HTTPS        |       ✅       |       ❌       |
+  | WS           |       ✅       |       ✅       |
+  
+  | Routes                                  | Tendermint-Go | Tendermint-Rs |
+  |-----------------------------------------|:-------------:|:-------------:|
+  | [Subscribe](#subscribe)                 |       ✅       |       ✅       |
+  | [Unsubscribe](#unsubscribe)             |       ✅       |       ✅       |
+  | [UnsubscribeAll](#unsubscribeall)       |       ✅       |       ❌       |
```
Comment (cmwaters) [inline spec/rpc/README.md:18]: These links seem to go nowhere - did you plan on having a link for them

File: spec/rpc/README.md [hunk 3]
```diff
@@ -0,0 +1,1267 @@
...
+  | [BlockResults](#blockresults)           |       ✅       |       ✅       |
+  | [Commit](#commit)                       |       ✅       |       ✅       |
+  | [Validators](#validators)               |       ✅       |       ✅       |
+  | [Genesis](#genesis)                     |       ✅       |       ✅       |
+  | [ConsensusParams](#consensusparams)     |       ✅       |       ❌       |
+  | [UnconfirmedTxs](#unconfirmedtxs)       |       ✅       |       ❌       |
+  | [NumUnconfirmedTxs](#numunconfirmedtxs) |       ✅       |       ❌       |
+  | [Tx](#tx)                               |       ✅       |       ❌       |
+  | [BroadCastTxSync](#broadcasttxsync)     |       ✅       |       ✅       |
+  | [BroadCastTxAsync](#broadcasttxasync)   |       ✅       |       ✅       |
```
Comment (cmwaters) [inline spec/rpc/README.md:34]: Are we getting rid off / no longer supporting `BroadcastCommit`

File: spec/rpc/README.md [hunk 4]
```diff
@@ -0,0 +1,1267 @@
+# RPC spec
+
+This file defines the JSON-RPC spec of Tendermint. This is meant to be implemented by all clients.
+
+## Support
+
+  |              | Tendermint-Go | Tendermint-Rs |
```
Comment (cmwaters) [inline spec/rpc/README.md:7]: nit: maybe add links to the repos here

**Review by cmwaters — COMMENTED** [review]
I know this is still a draft. But looking good so far.

A couple of extra points:
1. I think this could also be a good area to define the standard Tendermint RPC errors
2. We don't fully comply with the JSON RPC2.0 spec right? Here might be a good place to outline where exactly we diverge

File: spec/rpc/README.md (hunk 2 above)
Comment (marbar3778) [reply spec/rpc/README.md:18]: removed these

**Comment by marbar3778** [comment]
>     1. I think this could also be a good area to define the standard Tendermint RPC errors

Agree here, should this be a separate PR?

>     2. We don't fully comply with the JSON RPC2.0 spec right? Here might be a good place to outline where exactly we diverge

Instead of trying to define what we support of JSON RPC 2.0 and what we don't. I would like to say the spec is JSON RPC 2.0 compliant and Tendermint-Go is not adhering to the spec.  This will get  the Tendermint team to become spec compliant at the RPC layer.

**Comment by cmwaters** [comment]
> Agree here, should this be a separate PR?

Sure, up to you

File: spec/rpc/README.md [hunk 5]
```diff
@@ -0,0 +1,1264 @@
...
+        "remote_ip": "95.179.155.35"
+      }
+    ]
+  }
+}
+```
+
+### Blockchain
+
+Get block headers (max: 20) for minHeight <= height <= maxHeight.
```
Comment (cmwaters) [inline spec/rpc/README.md:245]: what does `height` mean here?

File: spec/rpc/README.md [hunk 6]
```diff
@@ -0,0 +1,1264 @@
...
+        }
+      ],
+      "app_hash": "",
+      "app_state": {}
+    }
+  }
+}
+```
+
+### ConsensusParams
```
Comment (cmwaters) [inline spec/rpc/README.md:844]: Doesn't this include `height` as a possible parameter.

File: spec/rpc/README.md [hunk 7]
```diff
@@ -0,0 +1,1264 @@
...
+}
+```
+
+### UnconfirmedTxs
+
+Get a list of unconfirmed transactions.
+
+#### Parameters
+
+- `limit (integer)` The amount of txs to respond with.
```
Comment (cmwaters) [inline spec/rpc/README.md:895]: I feel like this should be paginated or something. In tendermint-go this has a upper bound of 100 i.e. you can only request the top 100 txs

File: spec/rpc/README.md [hunk 8]
```diff
@@ -0,0 +1,1264 @@
...
+##### HTTP
+
+```sh
+curl  http://127.0.0.1:26657/unconfirmed_txs
+```
+
+##### JSONRPC
+
+```sh
+curl -X POST https://localhost:26657 -d "{\"jsonrpc\":\"2.0\",\"id\":1,\"method\":\"unconfirmed_txs\, \"params\":{\"limit\":\"20\"}"}"
```
Comment (cmwaters) [inline spec/rpc/README.md:908]:
```suggestion
curl -X POST https://localhost:26657 -d "{\"jsonrpc\":\"2.0\",\"id\":1,\"method\":\"unconfirmed_txs\, \"params\":{\"limit\":\"20\"}}"
```

File: spec/rpc/README.md (hunk 7 above)
Comment (marbar3778) [reply spec/rpc/README.md:895]: agree, it would be good to paginate more things.

File: spec/rpc/README.md [hunk 9]
```diff
@@ -0,0 +1,1268 @@
...
+  | [Genesis](#genesis)                     |                             ✅                              |                                 ✅                                 |
+  | [ConsensusParams](#consensusparams)     |                             ✅                              |                                 ❌                                 |
+  | [UnconfirmedTxs](#unconfirmedtxs)       |                             ✅                              |                                 ❌                                 |
+  | [NumUnconfirmedTxs](#numunconfirmedtxs) |                             ✅                              |                                 ❌                                 |
+  | [Tx](#tx)                               |                             ✅                              |                                 ❌                                 |
+  | [BroadCastTxSync](#broadcasttxsync)     |                             ✅                              |                                 ✅                                 |
```
Comment (cmwaters) [inline spec/rpc/README.md:30]: Why `BroadCast` instead of `Broadcast`?

File: spec/rpc/README.md [hunk 10]
```diff
@@ -0,0 +1,1268 @@
...
+        "remote_ip": "95.179.155.35"
+      }
+    ]
+  }
+}
+```
+
+### Blockchain
+
+Get block headers.
```
Comment (cmwaters) [inline spec/rpc/README.md:245]:
```suggestion
Get block headers. Returned in descending order. May be limited in quantity.
```

File: spec/rpc/README.md [hunk 11]
```diff
@@ -0,0 +1,1268 @@
...
+}
+```
+
+### ConsensusParams
+
+Get the consensus parameters.
+
+#### Parameters
+
+- `height (integer)`: Bloch height at which the consensus params would like to be fetched for.
```
Comment (cmwaters) [inline spec/rpc/README.md:850]:
```suggestion
- `height (integer)`: Block height at which the consensus params would like to be fetched for. 
```

**Review by cmwaters — APPROVED** [review]
Just a few minor things but otherwise good to go

**Pull request merged**